### PR TITLE
FlopCounterMode: Decompose ops for inference mode

### LINKED ISSUE #126268: ```FlopCounterMode``` returns 0 when inference mode is on during forwardpropagation.
### 🐛 Describe the bug

```FlopCounterMode``` reports 0 FLOPS when the forward propagation has the inference mode enabled by ```torch.inference_mode()```. The counter works fine if I use ```torch_no_grad()``` instead. The documentation says ```torch.inference_mode()``` disables the Forward-mode AD while ```torch.no_grad()``` does not. Maybe ```FlopCounterMode``` somehow relies on Forward-mode AD to calculate the computation cost?

PyTorch: 2.3.0
Python: 3.11.9

Test script 1 uses ```torch.inference_mode()```
```python
import torchvision.models as models
import torch
from torch.utils.flop_counter import FlopCounterMode

device = 'cuda'

inp = torch.randn(1, 3, 224, 224, device = device)
mod = models.resnet18().to(device)

@torch.inference_mode(True)
def run_batch(x):
    flop_counter = FlopCounterMode(mod)
    with flop_counter:
        return mod(x).sum()

run_batch(inp)
```

Output:
```
Module      FLOP    % Total
--------  ------  ---------
Global         0         0%
```

Test script 2 uses ```torch.no_grad()```
```python
import torchvision.models as models
import torch
from torch.utils.flop_counter import FlopCounterMode

device = 'cuda'

inp = torch.randn(1, 3, 224, 224, device = device)
mod = models.resnet18().to(device)

@torch.no_grad()
def run_batch(x):
    flop_counter = FlopCounterMode(mod)
    with flop_counter:
        return mod(x).sum()

run_batch(inp)
```

Output:
```
Module                     FLOP    % Total
--------------------  ---------  ---------
ResNet                3628.147M    100.00%
 - aten.convolution   3627.123M     99.97%
 - aten.addmm            1.024M      0.03%
 ResNet.conv1          236.028M      6.51%
  - aten.convolution   236.028M      6.51%
 ResNet.layer1         924.844M     25.49%
  - aten.convolution   924.844M     25.49%
 ResNet.layer2         822.084M     22.66%
  - aten.convolution   822.084M     22.66%
 ResNet.layer3         822.084M     22.66%
  - aten.convolution   822.084M     22.66%
 ResNet.layer4         822.084M     22.66%
  - aten.convolution   822.084M     22.66%
 ResNet.fc               1.024M      0.03%
  - aten.addmm           1.024M      0.03%
```

Test script 3 keeps gradient calculation enabled.
```python
import torchvision.models as models
import torch
from torch.utils.flop_counter import FlopCounterMode

device = 'cuda'

inp = torch.randn(1, 3, 224, 224, device = device)
mod = models.resnet18().to(device)

def run_batch(x):
    flop_counter = FlopCounterMode(mod)
    with flop_counter:
        return mod(x).sum()

run_batch(inp)
```

Output:
```
Module                     FLOP    % Total
--------------------  ---------  ---------
ResNet                3628.147M    100.00%
 - aten.convolution   3627.123M     99.97%
 - aten.addmm            1.024M      0.03%
 ResNet.conv1          236.028M      6.51%
  - aten.convolution   236.028M      6.51%
 ResNet.layer1         924.844M     25.49%
  - aten.convolution   924.844M     25.49%
 ResNet.layer2         822.084M     22.66%
  - aten.convolution   822.084M     22.66%
 ResNet.layer3         822.084M     22.66%
  - aten.convolution   822.084M     22.66%
 ResNet.layer4         822.084M     22.66%
  - aten.convolution   822.084M     22.66%
 ResNet.fc               1.024M      0.03%
  - aten.addmm           1.024M      0.03%
```

### Versions

```
PyTorch version: 2.3.0
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Red Hat Enterprise Linux Server release 7.9 (Maipo) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
Clang version: 3.4.2 (tags/RELEASE_34/dot2-final)
CMake version: version 2.8.12.2
Libc version: glibc-2.17

Python version: 3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:36:13) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.118.1.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla V100-PCIE-16GB
Nvidia driver version: 550.54.15
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                28
On-line CPU(s) list:   0-27
Thread(s) per core:    2
Core(s) per socket:    14
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Gold 6132 CPU @ 2.60GHz
Stepping:              4
CPU MHz:               2949.121
CPU max MHz:           3700.0000
CPU min MHz:           1000.0000
BogoMIPS:              5200.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              19712K
NUMA node0 CPU(s):     0-27
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba rsb_ctxsw ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_pkg_req pku ospke md_clear spec_ctrl intel_stibp flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.3.0
[pip3] torchaudio==2.3.0
[pip3] torchdiffeq==0.2.2
[pip3] torchvision==0.18.0
[pip3] triton==2.3.0
[conda] blas                      1.0                         mkl    conda-forge
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libblas                   3.9.0            16_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            16_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            16_linux64_mkl    conda-forge
[conda] mkl                       2022.2.1         h84fe81f_16997    conda-forge
[conda] numpy                     1.26.4          py311h64a7726_0    conda-forge
[conda] pytorch                   2.3.0           py3.11_cuda11.8_cudnn8.7.0_0    pytorch
[conda] pytorch-cuda              11.8                 h7e8668a_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.3.0               py311_cu118    pytorch
[conda] torchdiffeq               0.2.2              pyhd8ed1ab_0    conda-forge
[conda] torchtriton               2.3.0                     py311    pytorch
[conda] torchvision               0.18.0              py311_cu118    pytorch

```

cc @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

### PR DESCRIPTION
**Pull request by Feuermagier**
Fixes #126268

I've basically followed @ezyang suggestion (I think) to use `func.decompose(...)`. Since `__torch_dispatch__` won't be called a second time for the same op, I've added a second `TorchDispatchMode` (`_DecomposedCounterMode`) that simpy dispatches to the parent flop counter. Using `self` as the inner context manager is not possible, since the second call to `__enter__` would re-initialize the counter's tracking state.

Let me know if there's something wrong with this implementation, since I'm quite unsure how the decomposition thing actually works :D

### CODE REVIEW

File: torch/utils/flop_counter.py [hunk 1]
```diff
@@ -722,8 +723,12 @@ def __exit__(self, *args):
 
     def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         kwargs = kwargs if kwargs else {}
-        out = func(*args, **kwargs)
-        return self._count_flops(func._overloadpacket, out, args, kwargs)
+        if func._can_decompose():
+            with self.decomposed_counter:
```
Comment (ezyang) [inline torch/utils/flop_counter.py:727]: I think you don't need a separate dedicated mode, and you can just say `with self` here?

**Comment by ezyang** [comment]
Nice work! Let's see if it passes CI. Check if my comment makes sense

File: torch/utils/flop_counter.py (hunk 1 above)
Comment (Feuermagier) [reply torch/utils/flop_counter.py:727]: The issue is that entering `self` a second time resets the counter's state. There are two possible solutions:
- A second context manager that does not reset the counter's state, as my code does
- A flag that is set before re-entering the context manager, which is checked in `__enter__` to not reset the state

I'm not sure what's the better solution. What do you think?

**Comment by Feuermagier** [comment]
I'm still investigating the CI failure. For some reason, and only when using SDPA on nested tensors, `__torch_dispatch__` is called with ops like `aten::dim`, which the dispatcher doesn't recognize. Do you have any idea what the reason could be? This happens regardless of whether inference mode is active or not. Maybe it's due to the redispatch `NestedTensor` does in `__torch_dispatch__`.

File: torch/utils/flop_counter.py (hunk 1 above)
Comment (ezyang) [reply torch/utils/flop_counter.py:727]: Oh oops, I see the reset. The flopcounter shouldn't have been implemented this way. But yes, we will need an inner manager. Actually, if you're feeling adventurous, you could refactor the *entire* dispatch mode into its own dedicated private class, I'd like that better than duping it as is here

**Comment by ezyang** [comment]
@jbschlosser the NJT thing is just because of special size overload stuff right?

**Comment by Feuermagier** [comment]
As far as I understand, the root cause is that PyTorch calls __torch_dispatch__ of TorchDispatchModes before calling __torch_dispatch__ for tensor subclasses (what NJT is, and where the overloaded op is resolved). I think there are two workarounds:
- Check if an op is overridden by NJT; if so, skip the op (by returning NotImplemented from __torch_dispatch__). This breaks if there is any other tensor subclass that implements custom ops
- Check if `torch._C._dispatch_has_kernel_for_dispatch_key` throws a RuntimeError; if so, skip. This feels very hacky, but works for any custom tensor subclass & custom op

I've implemented the second option because I think it is better; let me know what you think @ezyang

**Comment by jbschlosser** [comment]
NJT is doing a fair amount of special stuff:
* It uses `dispatch_sizes` / `dispatch_layout` to get python-side `sizes()` / `layout()` implementations
* SDPA backend selection is implemented via `__torch_function__` (since it's CompositeImplicit)
* It looks like we're solving a similar problem that this PR addresses (i.e. we also want to do decompositions when in `inference_mode()`) via reentrant dispatch in #134683

File: torch/utils/flop_counter.py [hunk 2]
```diff
@@ -722,8 +724,24 @@ def __exit__(self, *args):
 
     def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         kwargs = kwargs if kwargs else {}
-        out = func(*args, **kwargs)
-        return self._count_flops(func._overloadpacket, out, args, kwargs)
+
+        dk = DispatchKey.CompositeImplicitAutograd
+        try:
+            can_dispatch = torch._C._dispatch_has_kernel_for_dispatch_key(func.name(), dk)
+        except RuntimeError:
```
Comment (jbschlosser) [inline torch/utils/flop_counter.py:731]: what conditions cause `_dispatch_has_kernel_for_dispatch_key` to throw?

**Comment by jbschlosser** [comment]
> As far as I understand, the root cause is that PyTorch calls torch_dispatch of TorchDispatchModes before calling torch_dispatch for tensor subclasses

This is true, yes.

> Check if an op is overridden by NJT; if so, skip the op (by returning NotImplemented from torch_dispatch). This breaks if there is any other tensor subclass that implements custom ops

I agree that any NJT-specific thing like this is **not** the right way to go. It's very possible other subclasses utilize `dispatch_sizes_strides_policy="sizes"` or ``dispatch_sizes_strides_policy="strides"``.

**Comment by jbschlosser** [comment]
For posterity, the error comes from a call to `findOp()` on the dispatcher, which is not aware of the special "ops" (e.g. `aten.dim.default`) used for a non-standard subclass `dispatch_sizes_strides_policy`:
https://github.com/pytorch/pytorch/blob/4dd04db5d0962fb8a1925cee831b9eb4f41ca4fd/torch/csrc/utils/python_dispatch.cpp#L581-L583

The `aten.dim.default` dispatch comes from:
https://github.com/pytorch/pytorch/blob/4dd04db5d0962fb8a1925cee831b9eb4f41ca4fd/c10/core/TensorImpl.cpp#L384-L386

Pretty sure there's no way `Dispatcher.findOp()` can access the `PyInterpreter` associated with the subclass instance's `PyObject` so I don't think it's feasible to put a dispatcher fix in to find the op.

A less hacky solution than catching a `RuntimeError` could be to put in special cases for the (small number of) size-related ops we expect for a non-standard `dispatch_sizes_strides_policy` setting. As done now, we'd want to return `NotImplemented` for all of these to allow the subclass instance to handle them.

File: torch/utils/flop_counter.py [hunk 3]
```diff
@@ -722,8 +724,34 @@ def __exit__(self, *args):
...
+            torch.ops.aten.size.default,
+            torch.ops.aten.sym_size.default,
+            torch.ops.aten.stride.default,
+            torch.ops.aten.sym_stride.default,
+            torch.ops.aten.storage_offset.default,
+            torch.ops.aten.sym_storage_offset.default,
+            torch.ops.aten.numel.default,
+            torch.ops.aten.sym_numel.default,
+            torch.ops.aten.dim.default, 
+            torch.ops.prim.layout.default]:
```
Comment (ezyang) [inline torch/utils/flop_counter.py:742]: this list is long enough that you should consider shoving it in a set and memoizing it

File: torch/utils/flop_counter.py [hunk 4]
```diff
@@ -722,8 +724,34 @@ def __exit__(self, *args):
 
     def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         kwargs = kwargs if kwargs else {}
-        out = func(*args, **kwargs)
-        return self._count_flops(func._overloadpacket, out, args, kwargs)
+
+        # Skip non-decomposoable ops from non-standard dispatch_sizes_strides_policy such as NJT
```
Comment (ezyang) [inline torch/utils/flop_counter.py:728]: decomposable

**Comment by ezyang** [comment]
this seems fine. waiting on ci

File: torch/utils/flop_counter.py (hunk 3 above)
Comment (Feuermagier) [reply torch/utils/flop_counter.py:742]: @jbschlosser is anything missing from the list? The values come from https://github.com/pytorch/pytorch/blob/b9f0563aafe9e21e71f5da6d7a6ffd62d9403641/torch/nested/_internal/ops.py#L414:L425 + `prim::layout` which was needed to pass the tests on my machine.

**Review by ezyang — DISMISSED** [review]

File: torch/utils/flop_counter.py (hunk 3 above)
Comment (jbschlosser) [reply torch/utils/flop_counter.py]: LGTM, thanks!

**Comment by jamesjwu** [comment]
Sorry, this PR is causing some issues in our internal jobs and it's bisecting to this:
```
AttributeError: 'ExternKernelSchedulerNode' object has no attribute '__get_estimated_runtime_cache'
```
Will see if we can get a repro later, but reverting for now

**Comment by ezyang** [comment]
aight taking a look

**Comment by ezyang** [comment]
internal xref: https://fb.workplace.com/groups/1075192433118967/permalink/1545856962719176/

**Comment by ezyang** [comment]
I think this fixes it

```
diff --git a/fbcode/caffe2/torch/utils/flop_counter.py b/fbcode/caffe2/torch/utils/flop_counter.py
--- a/fbcode/caffe2/torch/utils/flop_counter.py
+++ b/fbcode/caffe2/torch/utils/flop_counter.py
@@ -743,15 +743,16 @@
 
             return NotImplemented
 
-        dk = DispatchKey.CompositeImplicitAutograd
-        if torch._C._dispatch_has_kernel_for_dispatch_key(func.name(), dk):
-            # func can be decomposed; redispatch
+        # If we don't have func in flop_registry, see if it can decompose
+        if func not in self.flop_registry:
             with self.decomposed_counter:
-                return func._op_dk(dk, *args, **kwargs)
-        else:
-            # no further decomposition; execute & count flops
-            out = func(*args, **kwargs)
-            return self._count_flops(func._overloadpacket, out, args, kwargs)
+                r = func.decompose(*args, **kwargs)
+                if r is not NotImplemented:
+                    return r
+
+        # no further decomposition; execute & count flops
+        out = func(*args, **kwargs)
+        return self._count_flops(func._overloadpacket, out, args, kwargs)
 
     def _count_flops(self, func_packet, out, args, kwargs):
         if func_packet in self.flop_registry:
```

**Comment by ezyang** [comment]
TBH, I am not really sure how to test it though...

**Comment by ezyang** [comment]
The new version appears to still trigger some sort of Kineto problem

**Comment by ezyang** [comment]
With this commit it appears to pass internally too

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**